doc/net2disk.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH net2disk 1 "August 2011" Linux "User Manuals"

.SH NAME
.B net2disk
\- Capture data from network interface and record to disk.

.SH SYNOPSIS
.B net2disk
.B [--snaplen 
.I <capture length>
.B ]
.B [--promiscuous]
.B [--log_config
.I <filename>
.B ]
.B [--interfaces
.I <list of net interfaces>
.B ]
.B [--capture_files 
.I <list of files>
.B ]
.B [--smp_affinities
.I <list of affinities>
.B ]
.B [--ring_buffers
.I <total # ring buffers>
.B ]
.B [--write_blocks
.I <# write blocks per capture file>
.B ]
.B [--help]
.B ...

.SH DESCRIPTION
.B net2disk
Captures data from one or more network interfaces and writes data
to one or more disks. Each network interface is mapped to its own
disk. Each network interface/disk pair is processed in a separate 
process.

.SH OPTIONS
.IP "--snaplen <capture length>"
Length of bytes to capture (includes L2 header and above in the
protocol stack)

.IP "--promiscuous"
When present, sets each opened ethernet interface to promiscuous mode.

.IP "--log_config <filename>"
Configuration file for application logging subsystem.


.IP "--interfaces <list of net interfaces>"
List of network interfaces from which data are captured.

.IP "--capture_files <list of files>"
List of files to which data are written. These files are paired with
the network interfaces supplied in the --interfaces option, according
to their ordering on the command line.

.IP "--smp_affinities <list of affinities>"
List of SMP processor affinities. Each affinity is specified as a
number which is the number of the CPU to which a process should be
mapped (starting from 0). The arguments to the --interfaces,
--capture_files, and --smp_affinities options create a set of three
tuples fo the form: <interfaceN, fileN, affinityN>. This represents a
configuration where interfaceN and fileN are served in the same
process which is executed on processor affinityN.

.IP "--ring_buffers <total # of ring buffers>"
The total number of ring buffers to use in the system.

.IP "--write_blocks <# write blocks per capture file>"
The total number of write blocks to use for each capture file.

.IP"--help"
Print a usage message.


.SH FILES
.I /opt/haystack/mark6/etc/net2disk.conf
.RS
Application logging configuration file.

.RE
.I /opt/haystack/mark6/bin/run.sh
.RS
Execution script. This script will start a new net2disk instance.

.RE
.I /opt/haystack/mark6/var/net2disk.log
.RS
Log file. All application logging (DEBUG, INFO, WARNING, ERROR, etc.)
is sent to this log file.

.SH ENVIRONMENT
None

.SH DIAGNOSTICS
The following diagnostics may be issued on stderr:
 
Bad magic number.
.RS
The input file does not look like an archive file.
.RE
Old style baz segments.
.RS
.B foo
can only handle new style baz segments. COBOL
object libraries are not supported in this version.

.SH BUGS
None known at this time.

.SH AUTHOR
David Lapsley <del at haystack dot mit dot edu>

.SH "SEE ALSO"
.BR mark6 (1),
.BR dimino6 (1)



NEWS
Release 0.1	7/5/2011	David Lapsley<del@haystack.mit.edu>
-------------------------------------------------------------------

DESCRIPTION

First internal release. End to end data path, supporting infrastructure.

FEATURES

- Incorporated PF_RING code and module to get 8 Gbps NIC to Drive throughput.
- Implemented continuous integration and unit testing framework.
- Main program with options parsing.
- Support modules including:
	- BufferManager
	- ThreadPool
	- SocketManager
	- WriterTask

Release 0.2	7/7/2011	David Lapsley<del@haystack.mit.edu>
-------------------------------------------------------------------

DESCRIPTION

Major performance milestone.

Using a a single 10 GE network interface, and a single disk pack 
(with only 7 disks, one has failed), able to get ~ 7.6 Gbps net
to disk throughput (with 3% packet loss and ~95% utilization of a
single CPU -- out of 16 total CPUs). Throughput on both transmit
and receive checked, and volume of data written over test interval
checked. Test duraiton was 30 seconds.

FEATURES

- Ported code to debian squeeze (6.0.2) with 2.6.32 kernel.
- Removed source dependencies in favor of binary debian packages.
- Incorporated PF_RING module source into build tree.
- Upgraded "setup" build script. Includes more user accessible
  functionality. Also automatically installs all packages required
  to create a build/development machine from scratch.
- net2disk module now fully based on PF_RING architecture.
- Added tools subdirectory with network and disk configuration scripts.
  Disk configuration scripts include optimizations that disable
  journaling on ext4 file systems and enable driver level throughput
  optimizations.

Release 0.4	7/15/2011	David Lapsley<del@haystack.mit.edu>
-------------------------------------------------------------------

DESCRIPTION

Major performance milestone. Application software achieves maximum
network to disk throughput supported by hardware platform and drivers:
14.1 Gbps sustained for 60 seconds (still room to optimize platform
configuration). This was using 4 x 8-disk modules. Each module
configured as a separate RAID-0 array.

FEATURES

- Finalized application architecture (multi-process architecture
  rather than mult-threaded)
- Optimized PF_RING driver parameters
- Optimized Myricom driver settings
- Optimized IRQ SMP processor affinity for LSI MegaRAID and 
  Myricom Ethernet devices.
- Optimized application process SMP processor affinity.
- Optimized application buffer sizes, and data path.
- Optimized network packet capture code.
- Optimized internal buffering code.
- Optimized disk writing code.
- Application includes the following major classes:
	NetReader:
		Reads data from PF_RING buffers and fills BufferPool
		allocated buffers with the data.
	BufferPool:
		Internal buffers optimally sized and aligned for
		Direct Disk IO. Includes optimized memory management.
	FileWriter:
		Manages memory to disk data transfer.
	StatsWriter:
		Instrumentation class. Periodically logs network and
		disk performance statistics.
- Interactive command line interface that allows user to launch and
  stop capture threads in a convenient manner.
- Startup scripts that automatically set MegaRAID, Myricom, and
  Application processor affinities at startup.
- Disk module conditioning scripts that create and configure RAID
  arrays, label and partition virtual drives, create EXT4 file systems,
  and optimize file system parameters.


src/net2disk/net2disk.c
/*
 *
 * (C) 2005-11 - Luca Deri <deri@ntop.org>
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * VLAN support courtesy of Vincent Magnin <vincent.magnin@ci.unil.ch>
 *
 */

#define _GNU_SOURCE
#include <signal.h>
#include <sched.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/mman.h>
#include <errno.h>
#include <sys/poll.h>
#include <netinet/in_systm.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <net/ethernet.h>     /* the L2 protocols */
#include <sys/time.h>
#include <time.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <monetary.h>
#include <locale.h>

#include "pfring.h"

/* Constants */
#define ALARM_SLEEP            1
#define DEFAULT_SNAPLEN        9000
#define MAX_NUM_THREADS        64
#define DEFAULT_DEVICE         "eth0"
#define NUMBER_OF_FILES       	7 

/* Globals */
int pages_per_buffer;
int page_size;
int buffer_size;
u_char* bufs[NUMBER_OF_FILES];
int fds[NUMBER_OF_FILES];
pfring  *pd;
int verbose = 0, num_threads = 1;
pfring_stat pfringStats;
pthread_rwlock_t statsLock;

static struct timeval startTime;
unsigned long long numPkts[MAX_NUM_THREADS] = { 0 }, numBytes[MAX_NUM_THREADS] = { 0 };
u_int8_t wait_for_packet = 1, do_shutdown = 0;

/* Forward declarations. */
void writer_task(int fd, u_char* buf, int buf_size);


/* *************************************** */
/*
 * The time difference in millisecond
 */
double delta_time (struct timeval * now,
		   struct timeval * before) {
  time_t delta_seconds;
  time_t delta_microseconds;

  /*
   * compute delta in second, 1/10's and 1/1000's second units
   */
  delta_seconds      = now -> tv_sec  - before -> tv_sec;
  delta_microseconds = now -> tv_usec - before -> tv_usec;

  if(delta_microseconds < 0) {
    /* manually carry a one from the seconds field */
    delta_microseconds += 1000000;  /* 1e6 */
    -- delta_seconds;
  }
  return((double)(delta_seconds * 1000) + (double)delta_microseconds/1000);
}

/* ******************************** */
void print_stats() {
  pfring_stat pfringStat;
  struct timeval endTime;
  double deltaMillisec;
  static u_int8_t print_all;
  static u_int64_t lastPkts = 0;
  u_int64_t diff;
  static struct timeval lastTime;
  char buf1[64], buf2[64];

  if(startTime.tv_sec == 0) {
    gettimeofday(&startTime, NULL);
    print_all = 0;
  } else
    print_all = 1;

  gettimeofday(&endTime, NULL);
  deltaMillisec = delta_time(&endTime, &startTime);

  if(pfring_stats(pd, &pfringStat) >= 0) {
    double thpt;
    int i;
    unsigned long long nBytes = 0, nPkts = 0;

    for(i=0; i < num_threads; i++) {
      nBytes += numBytes[i];
      nPkts += numPkts[i];
    }

    thpt = ((double)8*nBytes)/(deltaMillisec*1000);

    fprintf(stderr, "=========================\n"
	    "Absolute Stats: [%u pkts rcvd][%u pkts dropped]\n"
	    "Total Pkts=%u/Dropped=%.1f %%\n",
	    (unsigned int)pfringStat.recv, (unsigned int)pfringStat.drop,
	    (unsigned int)(pfringStat.recv+pfringStat.drop),
	    pfringStat.recv == 0 ? 0 :
	    (double)(pfringStat.drop*100)/(double)(pfringStat.recv+pfringStat.drop));
    fprintf(stderr, "%s pkts - %s bytes", 
	    format_numbers((double)nPkts, buf1, sizeof(buf1), 0),
	    format_numbers((double)nBytes, buf2, sizeof(buf2), 0));

    if(print_all)
      fprintf(stderr, " [%s pkt/sec - %s Mbit/sec]\n",
	      format_numbers((double)(nPkts*1000)/deltaMillisec, buf1, sizeof(buf1), 1),
	      format_numbers(thpt, buf2, sizeof(buf2), 1));
    else
      fprintf(stderr, "\n");

    if(print_all && (lastTime.tv_sec > 0)) {
      deltaMillisec = delta_time(&endTime, &lastTime);
      diff = nPkts-lastPkts;
      fprintf(stderr, "=========================\n"
	      "Actual Stats: %llu pkts [%s ms][%s pkt/sec]\n",
	      (long long unsigned int)diff,
	      format_numbers(deltaMillisec, buf1, sizeof(buf1), 1),
	      format_numbers(((double)diff/(double)(deltaMillisec/1000)),  buf2, sizeof(buf2), 1));
    }

    lastPkts = nPkts;
  }

  lastTime.tv_sec = endTime.tv_sec, lastTime.tv_usec = endTime.tv_usec;

  fprintf(stderr, "=========================\n\n");
}

/* ******************************** */
void sigproc(int sig) {
  static int called = 0;

  fprintf(stderr, "Leaving...\n");
  if(called) return; else called = 1;
  do_shutdown = 1;
  print_stats();

  if(num_threads == 1)
    pfring_close(pd);

  exit(0);
}

/* ******************************** */
void my_sigalarm(int sig) {
  print_stats();
  alarm(ALARM_SLEEP);
  signal(SIGALRM, my_sigalarm);
}

/* ****************************************************** */
static char hex[] = "0123456789ABCDEF";

char* etheraddr_string(const u_char *ep, char *buf) {
  u_int i, j;
  char *cp;

  cp = buf;
  if ((j = *ep >> 4) != 0)
    *cp++ = hex[j];
  else
    *cp++ = '0';

  *cp++ = hex[*ep++ & 0xf];

  for(i = 5; (int)--i >= 0;) {
    *cp++ = ':';
    if ((j = *ep >> 4) != 0)
      *cp++ = hex[j];
    else
      *cp++ = '0';

    *cp++ = hex[*ep++ & 0xf];
  }

  *cp = '\0';
  return (buf);
}

/* ****************************************************** */
/*
 * A faster replacement for inet_ntoa().
 */
char* _intoa(unsigned int addr, char* buf, u_short bufLen) {
  char *cp, *retStr;
  u_int byte;
  int n;

  cp = &buf[bufLen];
  *--cp = '\0';

  n = 4;
  do {
    byte = addr & 0xff;
    *--cp = byte % 10 + '0';
    byte /= 10;
    if (byte > 0) {
      *--cp = byte % 10 + '0';
      byte /= 10;
      if (byte > 0)
	*--cp = byte + '0';
    }
    *--cp = '.';
    addr >>= 8;
  } while (--n > 0);

  /* Convert the string to lowercase */
  retStr = (char*)(cp+1);

  return(retStr);
}

/* ************************************ */
char* intoa(unsigned int addr) {
  static char buf[sizeof "ff:ff:ff:ff:ff:ff:255.255.255.255"];

  return(_intoa(addr, buf, sizeof(buf)));
}

/* ************************************ */
inline char* in6toa(struct in6_addr addr6) {
  static char buf[sizeof "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"];

  snprintf(buf, sizeof(buf),
	   "%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x",
	   addr6.s6_addr[0], addr6.s6_addr[1], addr6.s6_addr[2],
	   addr6.s6_addr[3], addr6.s6_addr[4], addr6.s6_addr[5], addr6.s6_addr[6],
	   addr6.s6_addr[7], addr6.s6_addr[8], addr6.s6_addr[9], addr6.s6_addr[10],
	   addr6.s6_addr[11], addr6.s6_addr[12], addr6.s6_addr[13], addr6.s6_addr[14],
	   addr6.s6_addr[15]);

  return(buf);
}

/* ****************************************************** */
char* proto2str(u_short proto) {
  static char protoName[8];

  switch(proto) {
  case IPPROTO_TCP:  return("TCP");
  case IPPROTO_UDP:  return("UDP");
  case IPPROTO_ICMP: return("ICMP");
  default:
    snprintf(protoName, sizeof(protoName), "%d", proto);
    return(protoName);
  }
}

/* ****************************************************** */
static int32_t thiszone;

/* DAVE */
unsigned char mybuf[DEFAULT_SNAPLEN];
void dummyProcesssPacket(const struct pfring_pkthdr *h, const u_char *p, const u_char *user_bytes) {
  long threadId = (long)user_bytes;

  if(verbose) {
    struct ether_header ehdr;
    u_short eth_type, vlan_id;
    char buf1[32], buf2[32];
    struct ip ip;
    int s;
    uint nsec;

    if(h->ts.tv_sec == 0) {
      gettimeofday((struct timeval*)&h->ts, NULL);
      parse_pkt((u_char*)p, (struct pfring_pkthdr*)h);
    }

    s = (h->ts.tv_sec + thiszone) % 86400;
    nsec = h->extended_hdr.timestamp_ns % 1000;
    
    printf("%02d:%02d:%02d.%06u%03u ",
	   s / 3600, (s % 3600) / 60, s % 60,
	   (unsigned)h->ts.tv_usec, nsec);

#if 0
    for(i=0; i<32; i++) printf("%02X ", p[i]);
    printf("\n");
#endif

    if(h->extended_hdr.parsed_header_len > 0) {
      printf("[eth_type=0x%04X]", h->extended_hdr.parsed_pkt.eth_type);
      printf("[l3_proto=%u]", (unsigned int)h->extended_hdr.parsed_pkt.l3_proto);

      printf("[%s:%d -> ", (h->extended_hdr.parsed_pkt.eth_type == 0x86DD) ?
	     in6toa(h->extended_hdr.parsed_pkt.ipv6_src) : intoa(h->extended_hdr.parsed_pkt.ipv4_src),
	     h->extended_hdr.parsed_pkt.l4_src_port);
      printf("%s:%d] ", (h->extended_hdr.parsed_pkt.eth_type == 0x86DD) ?
	     in6toa(h->extended_hdr.parsed_pkt.ipv6_dst) : intoa(h->extended_hdr.parsed_pkt.ipv4_dst),
	     h->extended_hdr.parsed_pkt.l4_dst_port);

      printf("[%s -> %s] ",
	     etheraddr_string(h->extended_hdr.parsed_pkt.smac, buf1),
	     etheraddr_string(h->extended_hdr.parsed_pkt.dmac, buf2));
    }

    memcpy(&ehdr, p+h->extended_hdr.parsed_header_len, sizeof(struct ether_header));
    eth_type = ntohs(ehdr.ether_type);

    printf("[%s -> %s][eth_type=0x%04X] ",
	   etheraddr_string(ehdr.ether_shost, buf1),
	   etheraddr_string(ehdr.ether_dhost, buf2), eth_type);


    if(eth_type == 0x8100) {
      vlan_id = (p[14] & 15)*256 + p[15];
      eth_type = (p[16])*256 + p[17];
      printf("[vlan %u] ", vlan_id);
      p+=4;
    }

    if(eth_type == 0x0800) {
      memcpy(&ip, p+h->extended_hdr.parsed_header_len+sizeof(ehdr), sizeof(struct ip));
      printf("[%s]", proto2str(ip.ip_p));
      printf("[%s:%d ", intoa(ntohl(ip.ip_src.s_addr)), h->extended_hdr.parsed_pkt.l4_src_port);
      printf("-> %s:%d] ", intoa(ntohl(ip.ip_dst.s_addr)), h->extended_hdr.parsed_pkt.l4_dst_port);

      printf("[tos=%d][tcp_seq_num=%u][caplen=%d][len=%d][parsed_header_len=%d]"
	     "[eth_offset=%d][l3_offset=%d][l4_offset=%d][payload_offset=%d]\n",
	     h->extended_hdr.parsed_pkt.ipv4_tos, h->extended_hdr.parsed_pkt.tcp.seq_num,
	     h->caplen, h->len, h->extended_hdr.parsed_header_len,
	     h->extended_hdr.parsed_pkt.offset.eth_offset,
	     h->extended_hdr.parsed_pkt.offset.l3_offset,
	     h->extended_hdr.parsed_pkt.offset.l4_offset,
	     h->extended_hdr.parsed_pkt.offset.payload_offset);

    } else {
      if(eth_type == 0x0806)
	printf("[ARP]");
      else
	printf("[eth_type=0x%04X]", eth_type);

      printf("[caplen=%d][len=%d][parsed_header_len=%d]"
	     "[eth_offset=%d][l3_offset=%d][l4_offset=%d][payload_offset=%d]\n",
	     h->caplen, h->len, h->extended_hdr.parsed_header_len,
	     h->extended_hdr.parsed_pkt.offset.eth_offset,
	     h->extended_hdr.parsed_pkt.offset.l3_offset,
	     h->extended_hdr.parsed_pkt.offset.l4_offset,
	     h->extended_hdr.parsed_pkt.offset.payload_offset);
    }
  }

  /** DAVE **/
  memcpy(mybuf, p, DEFAULT_SNAPLEN);
  numPkts[threadId]++, numBytes[threadId] += h->len;
}

/* *************************************** */
int32_t gmt2local(time_t t) {
  int dt, dir;
  struct tm *gmt, *loc;
  struct tm sgmt;

  if (t == 0)
    t = time(NULL);
  gmt = &sgmt;
  *gmt = *gmtime(&t);
  loc = localtime(&t);
  dt = (loc->tm_hour - gmt->tm_hour) * 60 * 60 +
    (loc->tm_min - gmt->tm_min) * 60;

  /*
   * If the year or julian day is different, we span 00:00 GMT
   * and must add or subtract a day. Check the year first to
   * avoid problems when the julian day wraps.
   */
  dir = loc->tm_year - gmt->tm_year;
  if (dir == 0)
    dir = loc->tm_yday - gmt->tm_yday;
  dt += dir * 24 * 60 * 60;

  return (dt);
}

/* *************************************** */
void printHelp(void) {
  printf("pfcount\n(C) 2005-11 Deri Luca <deri@ntop.org>\n\n");
  printf("-h              Print this help\n");
  printf("-i <device>     Device name. Use device@channel for channels, and dna:ethX for DNA\n");
  printf("-n <threads>    Number of polling threads (default %d)\n", num_threads);
  printf("-c <cluster id> cluster id\n");
  printf("-e <direction>  0=RX+TX, 1=RX only, 2=TX only\n");
  printf("-s <string>     String to search on packets\n");
  printf("-l <len>        Capture length\n");
  printf("-g <core_id>    Bind this app to a code (only with -n 0)\n");
  printf("-w <watermark>  Watermark\n");
  printf("-p <poll wait>  Poll wait (msec)\n");
  printf("-b <cpu %%>      CPU pergentage priority (0-99)\n");
  printf("-a              Active packet wait\n");
  printf("-r              Rehash RSS packets\n");
  printf("-v              Verbose\n");
}

/* *************************************** */
/* Bind this thread to a specific core */
int bind2core(u_int core_id) {
  cpu_set_t cpuset;
  int s;

  CPU_ZERO(&cpuset);
  CPU_SET(core_id, &cpuset);
  if((s = pthread_setaffinity_np(pthread_self(), sizeof(cpu_set_t), &cpuset)) != 0) {
    printf("Error while binding to core %u: errno=%i\n", core_id, s);
    return(-1);
  } else {
    return(0);
  }
}

void dump_buf(const long thread_id, const u_char* buf) {
  int i;
  printf("%ld ", thread_id);
  for(i=0; i<32; i++)
    printf("%02X ", buf[i]);
  printf("\n");
}

/* *************************************** */
void* packet_consumer_thread(void* _id) {
  long thread_id = (long)_id;
  int fd = fds[thread_id];
  u_char* filebuf = bufs[thread_id];
  u_char* netbuf;
  u_int numCPU = sysconf( _SC_NPROCESSORS_ONLN );
  u_long core_id = thread_id % numCPU;
  struct pfring_pkthdr hdr;

  printf("packet_consumer_thread(%lu)\n", thread_id);

  if((num_threads > 1) && (numCPU > 1)) {
    if(bind2core(core_id) == 0)
      printf("Set thread %lu on core %lu/%u\n", thread_id, core_id, numCPU);
  }

  memset(&hdr, 0, sizeof(hdr));

  while(1) {
    if (do_shutdown)
      break;

    int bytes_left = buffer_size;
    int bytes_read = 0;
    while (bytes_left > 0) {
      if (pfring_recv(pd, &netbuf, 0, &hdr, wait_for_packet) > 0) {
	numPkts[thread_id]++, numBytes[thread_id] += hdr.len;
#if 0
	dump_buf(thread_id, netbuf);
#endif
	if (do_shutdown)
	  break;

	int buffer_free = buffer_size - bytes_read;
	if (buffer_free >= DEFAULT_SNAPLEN) {
	  /* Copy entire received packet. */
	  memcpy(filebuf + bytes_read, netbuf, DEFAULT_SNAPLEN);
	  bytes_left -= DEFAULT_SNAPLEN;
	  bytes_read += DEFAULT_SNAPLEN;
	} else {
	  /* Pad out rest of buffer then write. */
	  memset(filebuf + bytes_read, 0, buffer_free);
	  writer_task(fd, filebuf, buffer_size);

	  /* Reset. */
	  bytes_left = buffer_size;
	  bytes_read = 0;
	}
      } else {
	if(wait_for_packet == 0)
	  sched_yield();
      }
    }
  }

  return(NULL);
}

/* *************************************** */
void
writer_task(int fd, u_char* buf, int buf_size) {
  // Write buffer to disk.
  int bytes_left = buf_size;
  int bytes_written = 0;

#if 0
  dump_buf((long)fd, buf);
#endif

  while (bytes_left) {
    int nb = write(fd, buf + bytes_written, buf_size);
    if (nb > 0) {
      bytes_left -= nb;
      bytes_written += nb;
    } else {
      perror("Unable to write to disk.");
    }
  }
}

/* *************************************** */
void
setup() {
  pages_per_buffer = 256;
  page_size = getpagesize();
  buffer_size = pages_per_buffer * page_size;

  char FILE_PREFIX[] = "/mnt/disk";
  char *PATHS[NUMBER_OF_FILES];
  int PATH_LENGTH = 255;
  int i;

  for (i=0; i<NUMBER_OF_FILES; i++) {
    if (posix_memalign((void*)&bufs[i], page_size, buffer_size) < 0) {
      perror("Memalign failed.");
      exit(1);
    }
  }

  for (i=0; i<NUMBER_OF_FILES; i++) {
    PATHS[i] = (char*)malloc(PATH_LENGTH);
    snprintf(PATHS[i], PATH_LENGTH, "%s%d/test.m6", FILE_PREFIX, i);

    int fd = open(PATHS[i], O_WRONLY | O_CREAT | O_DIRECT, S_IRWXU);
    if (fd < 0) {
      printf("i==%d fd==%d\n", i, fd);
      perror("Unable to open file descriptor.");
      exit(1);
    } else {
      printf("fds[%d]=%d\n", i, fd);
      fds[i] = fd;
    }
  }

  printf("page_size: %d", page_size);
  printf("buffer_size: %d", buffer_size);
}

/* *************************************** */
int main(int argc, char* argv[]) {
  char *device = NULL, c, *string = NULL, buf[32];
  u_char mac_address[6];
  int promisc, snaplen = DEFAULT_SNAPLEN, rc;
  u_int clusterId = 0;
  int bind_core = -1;
  packet_direction direction = rx_and_tx_direction;
  u_int16_t watermark = 0, poll_duration = 0, cpu_percentage = 0, rehash_rss = 0;

  startTime.tv_sec = 0;
  thiszone = gmt2local(0);

  while((c = getopt(argc,argv,"hi:c:dl:vs:ae:n:w:p:b:rg:" /* "f:" */)) != '?') {
    if((c == 255) || (c == -1)) break;

    switch(c) {
    case 'h':
      printHelp();
      return(0);
      break;
    case 'a':
      wait_for_packet = 0;
      break;
    case 'e':
      switch(atoi(optarg)) {
      case rx_and_tx_direction:
      case rx_only_direction:
      case tx_only_direction:
	direction = atoi(optarg);
	break;
      }
      break;
    case 'c':
      clusterId = atoi(optarg);
      break;
    case 'l':
      snaplen = atoi(optarg);
      break;
    case 'i':
      device = strdup(optarg);
      break;
    case 'n':
      num_threads = atoi(optarg);
      break;
    case 'v':
      verbose = 1;
      break;
    case 's':
      string = strdup(optarg);
      break;
    case 'w':
      watermark = atoi(optarg);
      break;
    case 'b':
      cpu_percentage = atoi(optarg);
      break;
    case 'p':
      poll_duration = atoi(optarg);
      break;
    case 'r':
      rehash_rss = 1;
      break;
    case 'g':
      bind_core = atoi(optarg);
      break;
    }
  }

  setup();

  /* Start */
  
  if(verbose) watermark = 1;
  if(device == NULL) device = DEFAULT_DEVICE;
  if(num_threads > MAX_NUM_THREADS) num_threads = MAX_NUM_THREADS;

  /* hardcode: promisc=1, to_ms=500 */
  promisc = 1;

  if(num_threads > 0)
    pthread_rwlock_init(&statsLock, NULL);

  if(wait_for_packet && (cpu_percentage > 0)) {
    if(cpu_percentage > 99) cpu_percentage = 99;
    pfring_config(cpu_percentage);
  }

  pd = pfring_open(device, promisc,  snaplen, (num_threads > 1) ? 1 : 0);

  if(pd == NULL) {
    printf("pfring_open error (pf_ring not loaded or perhaps you use quick mode and have already a socket bound to %s ?)\n",
	   device);
    return(-1);
  } else {
    u_int32_t version;

    pfring_set_application_name(pd, "pfcount");
    pfring_version(pd, &version);

    printf("Using PF_RING v.%d.%d.%d\n",
	   (version & 0xFFFF0000) >> 16,
	   (version & 0x0000FF00) >> 8,
	   version & 0x000000FF);
  }

  if(pfring_get_bound_device_address(pd, mac_address) != 0)
    printf("pfring_get_bound_device_address() failed\n");
  else
    printf("Capturing from %s [%s]\n", device, etheraddr_string(mac_address, buf));

  printf("# Device RX channels: %d\n", pfring_get_num_rx_channels(pd));
  printf("# Polling threads:    %d\n", num_threads);

  if(clusterId > 0) {
    rc = pfring_set_cluster(pd, clusterId, cluster_round_robin);
    printf("pfring_set_cluster returned %d\n", rc);
  }

  if((rc = pfring_set_direction(pd, direction)) != 0)
    printf("pfring_set_direction returned [rc=%d][direction=%d]\n", rc, direction);

  if(watermark > 0) {
    if((rc = pfring_set_poll_watermark(pd, watermark)) != 0)
      printf("pfring_set_poll_watermark returned [rc=%d][watermark=%d]\n", rc, watermark);
  }

  if(rehash_rss)
    pfring_enable_rss_rehash(pd);

  if(poll_duration > 0)
    pfring_set_poll_duration(pd, poll_duration);

  signal(SIGINT, sigproc);
  signal(SIGTERM, sigproc);
  signal(SIGINT, sigproc);


  if(!verbose) {
    signal(SIGALRM, my_sigalarm);
    alarm(ALARM_SLEEP);
  }

  pfring_enable_ring(pd);

  if(num_threads > 1) {
    pthread_t my_thread;
    long i;

    for(i=1; i<num_threads; i++)
      pthread_create(&my_thread, NULL, packet_consumer_thread, (void*)i);

    bind_core = -1;
  }

  if(bind_core >= 0)
    bind2core(bind_core);

  // if(1) {
  // pfring_loop(pd, dummyProcesssPacket, (u_char*)NULL);
  // } else
  packet_consumer_thread(0);

  alarm(0);
  sleep(1);
  pfring_close(pd);

  return(0);
}
